File: layer3-go-alm/internal/alm/pool_manager.go
// Pool Manager for Layer 3 ALM
// Manages multiple ALM instances (pools) within a single socket server,
// allowing concurrent access to different pools for multi-tenant/multi-experiment scenarios.

package alm

import (
	"fmt"
	"log"
	"sync"

	"github.com/mfn/layer3_alm/internal/config"
	"github.com/mfn/layer3_alm/internal/persistence"
)

// PoolManager manages multiple ALM instances
type PoolManager struct {
	pools   map[string]*ALM
	mu      sync.RWMutex
	dataDir string
	config  *config.ALMConfig
}

// NewPoolManager creates a new pool manager
func NewPoolManager(dataDir string, almConfig *config.ALMConfig) *PoolManager {
	return &PoolManager{
		pools:   make(map[string]*ALM),
		dataDir: dataDir,
		config:  almConfig,
	}
}

// GetOrCreatePool gets or creates a pool by ID
func (pm *PoolManager) GetOrCreatePool(poolID string) (*ALM, error) {
	// Fast path: check if pool already exists (read lock)
	pm.mu.RLock()
	if pool, exists := pm.pools[poolID]; exists {
		pm.mu.RUnlock()
		return pool, nil
	}
	pm.mu.RUnlock()

	// Slow path: create new pool (write lock)
	pm.mu.Lock()
	defer pm.mu.Unlock()

	// Double-check after acquiring write lock
	if pool, exists := pm.pools[poolID]; exists {
		return pool, nil
	}

	// Create persistence config for this pool
	persistConfig := &persistence.Config{
		DataDir:              pm.dataDir,
		PoolID:               poolID,
		FsyncIntervalMs:      1000,
		SnapshotIntervalSecs: 300,
		AofBufferSize:        64 * 1024,
	}

	log.Printf("Creating new ALM pool: %s", poolID)

	// Create the ALM instance with persistence
	almInstance, err := NewALMWithPersistence(pm.config, persistConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create pool %s: %w", poolID, err)
	}

	pm.pools[poolID] = almInstance
	log.Printf("Pool %s ready (total pools: %d)", poolID, len(pm.pools))

	return almInstance, nil
}

// GetPool retrieves a pool by ID (without creating)
func (pm *PoolManager) GetPool(poolID string) (*ALM, bool) {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	pool, exists := pm.pools[poolID]
	return pool, exists
}

// ListPools returns all active pool IDs
func (pm *PoolManager) ListPools() []string {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	poolIDs := make([]string, 0, len(pm.pools))
	for poolID := range pm.pools {
		poolIDs = append(poolIDs, poolID)
	}
	return poolIDs
}

// PoolCount returns the number of active pools
func (pm *PoolManager) PoolCount() int {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	return len(pm.pools)
}

// Close closes all pools and releases resources
func (pm *PoolManager) Close() error {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	log.Printf("Closing PoolManager with %d pools...", len(pm.pools))

	for poolID, pool := range pm.pools {
		if err := pool.Close(); err != nil {
			log.Printf("Error closing pool %s: %v", poolID, err)
		}
	}

	pm.pools = make(map[string]*ALM)
	log.Println("PoolManager closed")

	return nil
}

File: layer2-rust-dsr/src/pool_manager.rs
///! Multi-Pool Manager for DSR
///!
///! Manages multiple DSR instances (pools) within a single socket server,
///! allowing concurrent access to different pools for multi-tenant/multi-experiment scenarios.

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
use tokio::sync::RwLock;
use anyhow::Result;

use crate::{DynamicSimilarityReservoir, DSRConfig, PersistenceConfig};

/// Manager for multiple DSR pools
pub struct PoolManager {
    pools: RwLock<HashMap<String, Arc<DynamicSimilarityReservoir>>>,
    data_dir: PathBuf,
    config: DSRConfig,
}

impl PoolManager {
    /// Create a new pool manager
    pub fn new(data_dir: PathBuf, config: DSRConfig) -> Self {
        Self {
            pools: RwLock::new(HashMap::new()),
            data_dir,
            config,
        }
    }

    /// Get or create a pool by ID
    pub async fn get_or_create_pool(&self, pool_id: &str) -> Result<Arc<DynamicSimilarityReservoir>> {
        // Fast path: check if pool already exists
        {
            let pools = self.pools.read().await;
            if let Some(pool) = pools.get(pool_id) {
                return Ok(Arc::clone(pool));
            }
        }

        // Slow path: create new pool
        let mut pools = self.pools.write().await;

        // Double-check after acquiring write lock
        if let Some(pool) = pools.get(pool_id) {
            return Ok(Arc::clone(pool));
        }

        // Create persistence config for this pool
        let persistence_config = PersistenceConfig {
            data_dir: self.data_dir.clone(),
            pool_id: pool_id.to_string(),
            fsync_interval_ms: 1000,
            snapshot_interval_secs: 300,
            aof_buffer_size: 64 * 1024,
        };

        tracing::info!("Creating new DSR pool: {}", pool_id);

        // Create pool with persistence
        let pool = if persistence_config.aof_path().exists() || persistence_config.snapshot_path().exists() {
            tracing::info!("Recovering pool {} from persistence", pool_id);
            Arc::new(DynamicSimilarityReservoir::recover_from_persistence(
                self.config.clone(),
                persistence_config,
            ).await?)
        } else {
            tracing::info!("Creating fresh pool {}", pool_id);
            Arc::new(DynamicSimilarityReservoir::new_with_persistence(
                self.config.clone(),
                Some(persistence_config),
            )?)
        };

        pools.insert(pool_id.to_string(), Arc::clone(&pool));
        tracing::info!("Pool {} ready (total pools: {})", pool_id, pools.len());

        Ok(pool)
    }

    /// Get pool if it exists (without creating)
    pub async fn get_pool(&self, pool_id: &str) -> Option<Arc<DynamicSimilarityReservoir>> {
        let pools = self.pools.read().await;
        pools.get(pool_id).map(Arc::clone)
    }

    /// List all active pool IDs
    pub async fn list_pools(&self) -> Vec<String> {
        let pools = self.pools.read().await;
        pools.keys().cloned().collect()
    }

    /// Get pool count
    pub async fn pool_count(&self) -> usize {
        let pools = self.pools.read().await;
        pools.len()
    }
}
